plan/rule_table_partition.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+// tablePartition rewrites the ast for table partition.
+//
+// create table t (id int) partition by range (id)
+//   (partition
+//      p1 values less than (10),
+//      p2 values less than (20),
+//      p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// tablePartition is here because it's easier to prune partition after predicate push down.
+type tablePartition struct{}
```

Comment:
s/tablePartition/partitionPrunning/?

Reply:
It not only does the partition pruning, but also the ast structure rewrite, so I think partitionPrunning is not very accurate

Reply:
s/tablePartition/partitionPrunner